### PR TITLE
Yaml support for Keras Sequential models

### PR DESCRIPTION
**Pull request by maxpumperla**
- Sequential now has to_yaml() method to dump models as yaml strings.
- models module has sequential_from_yaml() method to load from such strings
- Models can be stored with and without parameters.
- It's possible to (de-)serialize non-standard optimizers, constraints and regularizers.
- Includes a test that checks for equality of configs before and after serialization.

### CODE REVIEW

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -5,16 +5,15 @@
...
 from ..utils.theano_utils import shared_zeros
 from ..layers.core import Layer
 
-
 class Convolution1D(Layer):
     def __init__(self, input_dim, nb_filter, filter_length,
         init='uniform', activation='linear', weights=None,
         border_mode='valid', subsample_length=1,
-        W_regularizer=None, b_regularizer=None, activity_regularizer=None, W_constraint=None, b_constraint=None):
+        W_regularizer='identity', b_regularizer='identity', activity_regularizer='identity', W_constraint='identity', b_constraint='identity'):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: For efficiency reasons I believe it is preferable to keep the default regularizers to `None`. This is also important because using `None` in the class constructor is part of the standard API and the class should be able to deal with it.

This is absolutely not an issue in `get_config`: you can simply use, e.g.:
```python
"W_constraint":self.W_constraint.get_config() if  self.W_constraint else None,
```
After setting `self.W_constraint` to the value given to the constructor in `__init__`.

File: keras/models.py [hunk 2]
```diff
@@ -3,7 +3,19 @@
 import theano
 import theano.tensor as T
 import numpy as np
-import warnings, time, copy
+import warnings, time, copy, yaml
+
+from .layers.convolutional import *
```
Comment (fchollet) [inline keras/models.py]: Let's avoid `*` imports, even if the import list is very long. The namespace in `models` is very busy, and we need what is free and what isn't.

File: keras/models.py [hunk 3]
```diff
@@ -76,6 +88,49 @@ def standardize_weights(y, sample_weight=None, class_weight=None):
...
+    '''
+        Returns a compiled Sequential model generated from a local yaml file,
+        which is either created by hand or from Sequential.to_yaml
+    '''
+    modelYaml = yaml.load(yamlString)
+    model = Sequential()
+
+    class_mode = modelYaml.get('class_mode')
+    theano_mode = modelYaml.get('theano_mode')
+    loss = globals()[modelYaml.get('loss')]
```
Comment (fchollet) [inline keras/models.py]: This would not work with custom loss functions, I believe. It is also somewhat hack-ish and should be avoided.

In fact, why should we compile the model at this stage? Shouldn't we simply return a pre-compilation model, to be compiled by the user with whatever optimizer and loss they see fit?

File: keras/models.py [hunk 4]
```diff
@@ -237,6 +292,8 @@ class Sequential(Model, containers.Sequential):
 
     def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
         self.optimizer = optimizers.get(optimizer)
+
+        self.unweighted_loss = objectives.get(loss)
```
Comment (fchollet) [inline keras/models.py]: In which case this wouldn't be necessary anymore.

File: keras/models.py [hunk 5]
```diff
@@ -76,6 +88,49 @@ def standardize_weights(y, sample_weight=None, class_weight=None):
...
+    model = Sequential()
+
+    class_mode = modelYaml.get('class_mode')
+    theano_mode = modelYaml.get('theano_mode')
+    loss = globals()[modelYaml.get('loss')]
+
+    optim = modelYaml.get('optimizer')
+    optimName = optim.get('name')
+    optim.pop('name')
+    optimizer = globals()[optimName](**optim)
```
Comment (fchollet) [inline keras/models.py]: In general throughout this PR, we should follow PEP8 and avoid camel case. Let's use snake case instead (this is Python after all).

File: keras/models.py [hunk 6]
```diff
@@ -76,6 +88,49 @@ def standardize_weights(y, sample_weight=None, class_weight=None):
...
+        hasParams = False
+        if layer.has_key('parameters'):
+            params = layer.get('parameters')
+            layer.pop('parameters')
+            hasParams = True
+        for k, v in layer.iteritems():
+            if isinstance(v, dict):
+                vname = v.get('name')
+                v.pop('name')
+                layer[k] = globals()[vname](**v)
```
Comment (fchollet) [inline keras/models.py]: This seems dangerous. I would definitely prefer a more verbose, more explicit approach. First determine if the object is a constraint/regularizer/etc, then get it from the appropriate module.

**Comment by fchollet** [comment]
I believe this would not work with Sequential models that include one or several Merge layers. What is your take on that?

File: keras/layers/convolutional.py (hunk 1 above)
Comment (maxpumperla) [inline keras/layers/convolutional.py]: Completely agree with you, wasn't too happy either. I like your suggestion.

File: keras/models.py (hunk 3 above)
Comment (maxpumperla) [inline keras/models.py]: It's a bit of a hack, but very similar to
~~~~{.python}
from .utils.generic_utils import get_from_module
def get(identifier):
    return get_from_module(identifier, globals(), 'objective')
~~~~
used anywhere else.

It's a design choice, but I'm not at all adamant on keeping the compilation step. For me, there's two scenarios:
* Specify an untrained model w/o parameters, which does not have to be compiled.
* Train a model and store every bit of it, including optimizers, in case you want to pick it up later for further training etc. In this case it makes sense to have the compiled version at hand. At least I would like to have the option to have a compiled version.

File: keras/models.py (hunk 4 above)
Comment (maxpumperla) [inline keras/models.py]: Sure, we could role this back!

**Comment by maxpumperla** [comment]
I haven't thought about Merge layers much. In fact, I was under the impression that a Sequential shouldn't allow for that. For now, this PR is meant as serialisation of sequential (in its literal meaning) models. Merging Sequentials would need a description language of graphs, i.e. how to connect Sequentials. Right now I'm not sure how to do this properly. Any thoughts on your side?

File: keras/models.py (hunk 3 above)
Comment (fchollet) [inline keras/models.py]: It's the same hack, but the big difference is `get` is only used in small files where the set of names you want to be able to "get" is identical (or close enough) to the entire namespace. Which makes it safe. The reason this hack is potentially dangerous in the first place is the likelihood of namespace collisions, and they are highly likely in a file as busy as `models.py`. 

So I'd rather import each module separately, then used `getattr` on the appropriate module name. Then we limit the use of the namespace to the name of modules, which is fine.

Comment (fchollet) [inline keras/models.py]: > Train a model and store every bit of it, including optimizers, in case you want to pick it up later for further training etc. In this case it makes sense to have the compiled version at hand. At least I would like to have the option to have a compiled version.

Sounds very useful, but there are further problems with implementing this: for instance, how do you save the state of the optimizer for further training? That state is contained in shared Theano variables.

I have no strong opinion either way, the main argument I see for skipping compilation is the use of custom objectives (custom optimizers are relatively unlikely, much like custom regularizers and constraints, however custom objectives are fairly common).

**Comment by fchollet** [comment]
Actually I believe your code could be reused for serializing / loading multi-branch sequential models (and Graph models as well) very easily. First you identify the Merge layers, then for each Merge layer you serialize the merged containers, as a list. Recursively.

**Comment by maxpumperla** [comment]
I've tried to integrate all your suggestions in the latest update. Notably, we can now de-/serialize general Sequential and Graph models, which is also reflected in tests. The namespace of models.py is kept clean and it is possible to serialize both compiled and uncompiled models. With very little changes we could get rid of the compiled one, if you insist.

**Comment by fchollet** [comment]
Thank you for the update!

I merged it, with the following modifications:
- Style standardization
- Python3 compatibility (used a few Python2-only features) 
- Made `pyyaml` a dependency in `setup.py` (yaml support is not part of the Python stdlib)
- Removed weight saving in yaml files. 

The last point was motivated by the following reasons:
- Readability: the main point of yaml files for storing model configuration is that humans can read and edit them. Storing weights alongside the model structure would break this.
- Efficiency: we already have a very efficient way to save weights (HDF5), whereas dumping them to yaml would be quite inefficient both time-wise and space-wise. It is also always preferable to only have one obvious way to do something, not two. 
- Avoiding redundancy in the codebase: weight saving required a set of `layer_to_yaml` methods that were effectively the same as `get_config`, with weights. We should only have one `get_config`. Additionally, the naming was rather misleading since these methods did not in fact return a yaml object but instead a Python dictionary. Besides, the intuitive naming for these methods would have been `to_yaml` rather than `layer_to_yaml` (since they are methods of `Layer`/`Container` objects, therefore they obviously apply to a layer-like), which would then have caused recursion issues since they would have had a different behavior than `Model.to_yaml` (which does return a yaml string).

TL;DR: `layer_to_yaml` is gone, we're using `get_config` instead. Weight saving will be done in HDF5:
```Python
graph_yaml = graph.to_yaml()
graph.save_weights('temp.h5')

reloaded_graph = model_from_yaml(graph_yaml)
reloaded_graph.load_weights('temp.h5')
```

Thank you for the valuable contribution!

**Comment by lukedeo** [comment]
This fixes #376, so closing. This is awesome, thanks!
